File: apps/flutter/lib/features/transactions/domain/entities/transaction.dart
class Transaction {
  final String id;
  final String type;
  final DateTime date;
  final double amount;
  final String currency;
  final String? categoryId;
  final String? fromAccountId;
  final String? toAccountId;
  final String? note;
  final List<String> tags;
  final String status;
  final DateTime? clearedAt;
  final String? recurringRuleId;

  const Transaction({
    required this.id,
    required this.type,
    required this.date,
    required this.amount,
    required this.currency,
    required this.categoryId,
    required this.fromAccountId,
    required this.toAccountId,
    required this.note,
    required this.tags,
    required this.status,
    required this.clearedAt,
    this.recurringRuleId,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: (json["transactionId"] ?? json["id"]) as String,
      type: json["type"] as String,
      date: DateTime.parse(json["date"] as String),
      amount: (json["amount"] as num?)?.toDouble() ?? 0,
      currency: json["currency"] as String? ?? "BRL",
      categoryId: json["categoryId"] as String?,
      fromAccountId: json["fromAccountId"] as String?,
      toAccountId: json["toAccountId"] as String?,
      note: json["note"] as String?,
      tags:
          (json["tags"] as List<dynamic>?)
              ?.map((item) => item.toString())
              .toList() ??
          const [],
      status: json["status"] as String? ?? "pending",
      clearedAt: json["clearedAt"] != null
          ? DateTime.tryParse(json["clearedAt"] as String)
          : null,
      recurringRuleId: json["recurringRuleId"] as String?,
    );
  }
}

File: apps/api/src/models/transaction.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria";
import { createMongoId } from "./shared/mongo_id";

export enum TransactionType {
  Income = "income",
  Expense = "expense",
  Transfer = "transfer",
}

export type TransactionStatus = "pending" | "cleared";

export type TransactionPrimitives = {
  id?: string;
  transactionId: string;
  userId: string;
  type: TransactionType;
  date: Date;
  amount: number;
  currency: string;
  categoryId?: string | null;
  fromAccountId?: string | null;
  toAccountId?: string | null;
  note?: string | null;
  tags?: string[] | null;
  status: TransactionStatus;
  clearedAt?: Date | null;
  createdAt: Date;
  updatedAt?: Date;
  deletedAt?: Date | null;
  recurringRuleId?: string;
};

export type TransactionCreateProps = {
  userId: string;
  type: TransactionType;
  date: Date;
  amount: number;
  currency: string;
  categoryId?: string | null;
  fromAccountId?: string | null;
  toAccountId?: string | null;
  note?: string | null;
  tags?: string[] | null;
  status?: TransactionStatus;
  clearedAt?: Date | null;
  recurringRuleId?: string;
};

export class Transaction extends AggregateRoot {
  private readonly props: TransactionPrimitives;

  private constructor(props: TransactionPrimitives) {
    super();
    this.props = props;
  }

  static create(props: TransactionCreateProps): Transaction {
    const now = new Date();
    const status = props.status ?? "pending";
    const clearedAt = status === "cleared" ? props.clearedAt ?? now : null;

    return new Transaction({
      userId: props.userId,
      type: props.type,
      date: props.date,
      amount: props.amount,
      currency: props.currency,
      categoryId: props.categoryId ?? null,
      fromAccountId: props.fromAccountId ?? null,
      toAccountId: props.toAccountId ?? null,
      note: props.note ?? null,
      tags: props.tags ?? null,
      status,
      clearedAt,
      recurringRuleId: props.recurringRuleId,
      transactionId: createMongoId(),
      createdAt: now,
      updatedAt: now,
      deletedAt: null,
    });
  }

  getId(): string {
    return this.props.id ?? this.props.transactionId;
  }

  getTransactionId(): string {
    return this.props.transactionId;
  }

  toPrimitives(): TransactionPrimitives {
    return this.props;
  }

  static fromPrimitives(primitives: TransactionPrimitives): Transaction {
    return new Transaction(primitives);
  }
}
